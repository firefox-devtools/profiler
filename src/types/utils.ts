/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

/**
 * This type serves as documentation for how an array is meant to be used, but does
 * not support type checking. We often use an Array instead of a Map to translate
 * one type of index into another type of index. This is similar to how we use the
 * Map<K,V> type, but with the Array.
 */
export type IndexedArray<_IndexType, Value> = Array<Value>;

export type MixedObject = { [key: string]: unknown };
