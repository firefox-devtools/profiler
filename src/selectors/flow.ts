/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import { createSelector } from 'reselect';

import {
  computeProfileFlowInfo,
  computeFlowTiming,
} from '../profile-logic/marker-data';
import { getThreadSelectors } from './per-thread';
import { getActiveFlows } from './url-state';
import type { ThreadSelectors } from './per-thread';
import {
  getThreads,
  getMarkerSchema,
  getRawProfileSharedData,
} from './profile';

import type {
  Selector,
  State,
  MarkerIndex,
  Marker,
  ProfileFlowInfo,
  FlowTiming,
} from 'firefox-profiler/types';

function _arraysShallowEqual(arr1: any[], arr2: any[]): boolean {
  return arr1.length === arr2.length && arr1.every((val, i) => val === arr2[i]);
}

function _createSelectorForAllThreads<T>(
  f: (selectors: ThreadSelectors, state: State) => T
): Selector<T[]> {
  let previousOutputPerThread: T[] = [];
  return function recomputeSelectorForAllThreads(state: State): T[] {
    const threads = getThreads(state);
    let outputPerThread = threads.map((_thread, i) => {
      const threadSelectors = getThreadSelectors(i);
      return f(threadSelectors, state);
    });
    if (_arraysShallowEqual(outputPerThread, previousOutputPerThread)) {
      outputPerThread = previousOutputPerThread;
    }
    previousOutputPerThread = outputPerThread;
    return outputPerThread;
  };
}

export const getFullMarkerListPerThread: Selector<Marker[][]> =
  _createSelectorForAllThreads(({ getFullMarkerList }, state) =>
    getFullMarkerList(state)
  );

export const getMarkerChartLabelGetterPerThread: Selector<
  Array<(marker: MarkerIndex) => string>
> = _createSelectorForAllThreads(({ getMarkerChartLabelGetter }, state) =>
  getMarkerChartLabelGetter(state)
);

export const getProfileFlowInfo: Selector<ProfileFlowInfo> = createSelector(
  getFullMarkerListPerThread,
  getThreads,
  getMarkerSchema,
  getRawProfileSharedData,
  computeProfileFlowInfo
);

export const getFlowTiming: Selector<FlowTiming> = createSelector(
  getProfileFlowInfo,
  getActiveFlows,
  computeFlowTiming
);
