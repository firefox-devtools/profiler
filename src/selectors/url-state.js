/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import { createSelector } from 'reselect';
import { ensureExists, getFirstItemFromSet } from '../utils/flow';
import { urlFromState } from '../app-logic/url-handling';
import { getThreadsKey } from '../profile-logic/profile-data';
import { stringsToMarkerRegExps } from '../profile-logic/marker-data';
import { getProfileNameFromZipPath } from 'firefox-profiler/profile-logic/zip-files';
import { SYMBOL_SERVER_URL } from '../app-logic/constants';
import { splitSearchString, stringsToRegExp } from '../utils/string';
import { isLocalURL } from '../utils/url';

import type {
  ThreadIndex,
  Pid,
  TransformStack,
  Action,
  TimelineType,
  DataSource,
  ImplementationFilter,
  CallTreeSummaryStrategy,
  UrlState,
  TimelineTrackOrganization,
  Selector,
  DangerousSelectorWithArguments,
  StartEndRange,
  TrackIndex,
  ThreadsKey,
  ProfileSpecificUrlState,
  FullProfileSpecificUrlState,
  ActiveTabSpecificProfileUrlState,
  NativeSymbolInfo,
  TabID,
} from 'firefox-profiler/types';

import type { TabSlug } from '../app-logic/tabs-handling';
import type { MarkerRegExps } from '../profile-logic/marker-data';

import urlStateReducer from '../reducers/url-state';
import { formatMetaInfoString } from '../profile-logic/profile-metainfo';

/**
 * Various simple selectors into the UrlState.
 */
export const getUrlState: Selector<UrlState> = (state): UrlState =>
  state.urlState;
export const getProfileSpecificState: Selector<ProfileSpecificUrlState> = (
  state
) => getUrlState(state).profileSpecific;
export const getFullProfileSpecificState: Selector<
  FullProfileSpecificUrlState,
> = (state) => getProfileSpecificState(state).full;
export const getActiveTabProfileSpecificState: Selector<
  ActiveTabSpecificProfileUrlState,
> = (state) => getProfileSpecificState(state).activeTab;

export const getDataSource: Selector<DataSource> = (state) =>
  getUrlState(state).dataSource;
export const getHash: Selector<string> = (state) => getUrlState(state).hash;
export const getProfileUrl: Selector<string> = (state) =>
  getUrlState(state).profileUrl;
export const getProfilesToCompare: Selector<string[] | null> = (state) =>
  getUrlState(state).profilesToCompare;
export const getProfileNameFromUrl: Selector<string | null> = (state) =>
  getUrlState(state).profileName;
export const getAllCommittedRanges: Selector<StartEndRange[]> = (state) =>
  getProfileSpecificState(state).committedRanges;
export const getImplementationFilter: Selector<ImplementationFilter> = (
  state
) => getProfileSpecificState(state).implementation;
export const getLastSelectedCallTreeSummaryStrategy: Selector<
  CallTreeSummaryStrategy,
> = (state) =>
  getProfileSpecificState(state).lastSelectedCallTreeSummaryStrategy;
export const getShowUserTimings: Selector<boolean> = (state) =>
  getProfileSpecificState(state).showUserTimings;
export const getSourceViewFile: Selector<string | null> = (state) =>
  getProfileSpecificState(state).sourceView.sourceFile;
export const getSourceViewScrollGeneration: Selector<number> = (state) =>
  getProfileSpecificState(state).sourceView.scrollGeneration;
export const getAssemblyViewIsOpen: Selector<boolean> = (state) =>
  getProfileSpecificState(state).assemblyView.isOpen;
export const getAssemblyViewNativeSymbol: Selector<NativeSymbolInfo | null> = (
  state
) => getProfileSpecificState(state).assemblyView.nativeSymbol;
export const getAssemblyViewScrollGeneration: Selector<number> = (state) =>
  getProfileSpecificState(state).assemblyView.scrollGeneration;
export const getShowJsTracerSummary: Selector<boolean> = (state) =>
  getFullProfileSpecificState(state).showJsTracerSummary;
export const getTimelineTrackOrganization: Selector<
  TimelineTrackOrganization,
> = (state) => getUrlState(state).timelineTrackOrganization;

/**
 * Active tab specific url state selectors
 */
export const getIsActiveTabResourcesPanelOpen: Selector<boolean> = (state) =>
  getActiveTabProfileSpecificState(state).isResourcesPanelOpen;

/**
 * Raw search strings, before any splitting has been performed.
 */
export const getCurrentSearchString: Selector<string> = (state) =>
  getProfileSpecificState(state).callTreeSearchString;
export const getMarkersSearchString: Selector<string> = (state) =>
  getProfileSpecificState(state).markersSearchString;
export const getNetworkSearchString: Selector<string> = (state) =>
  getProfileSpecificState(state).networkSearchString;
export const getSelectedTab: Selector<TabSlug> = (state) =>
  getUrlState(state).selectedTab;
export const getInvertCallstack: Selector<boolean> = (state) =>
  getSelectedTab(state) === 'calltree' &&
  getProfileSpecificState(state).invertCallstack;

export const getSelectedThreadIndexesOrNull: Selector<
  Set<ThreadIndex> | null,
> = (state) => getProfileSpecificState(state).selectedThreads;
export const getSelectedThreadIndexes: Selector<Set<ThreadIndex>> = (state) =>
  ensureExists(
    getSelectedThreadIndexesOrNull(state),
    'Attempted to get a thread index before a profile was loaded.'
  );
export const getSelectedThreadsKey: Selector<ThreadsKey> = (state) =>
  getThreadsKey(getSelectedThreadIndexes(state));

/**
 * This selector is temporary for a migration to multiple selected thread indexes.
 */
export const getFirstSelectedThreadIndex: Selector<ThreadIndex> = (state) =>
  ensureExists(
    getFirstItemFromSet(getSelectedThreadIndexes(state)),
    'Expected to find at least one thread index in the selected thread indexes'
  );
export const getTimelineType: Selector<TimelineType> = (state) =>
  getProfileSpecificState(state).timelineType;

/**
 * Simple selectors for tracks and track order.
 */
export const getLegacyThreadOrder: Selector<ThreadIndex[] | null> = (state) =>
  getFullProfileSpecificState(state).legacyThreadOrder;
export const getLegacyHiddenThreads: Selector<ThreadIndex[] | null> = (state) =>
  getFullProfileSpecificState(state).legacyHiddenThreads;
export const getGlobalTrackOrder: Selector<TrackIndex[]> = (state) =>
  getFullProfileSpecificState(state).globalTrackOrder;
export const getHiddenGlobalTracks: Selector<Set<TrackIndex>> = (state) =>
  getFullProfileSpecificState(state).hiddenGlobalTracks;
export const getHiddenLocalTracksByPid: Selector<Map<Pid, Set<TrackIndex>>> = (
  state
) => getFullProfileSpecificState(state).hiddenLocalTracksByPid;
export const getLocalTrackOrderByPid: Selector<Map<Pid, TrackIndex[]>> = (
  state
) => getFullProfileSpecificState(state).localTrackOrderByPid;
export const getTabFilter: Selector<TabID | null> = (state) =>
  getFullProfileSpecificState(state).tabFilter;
export const hasTabFilter: Selector<boolean> = (state) =>
  getTabFilter(state) !== null;

/**
 * This selector does a simple lookup in the set of hidden tracks for a PID, and ensures
 * that a TrackIndex is selected correctly. This makes it easier to avoid doing null
 * checks everywhere.
 */
export const getHiddenLocalTracks: DangerousSelectorWithArguments<
  Set<TrackIndex>,
  Pid,
> = (state, pid) =>
  ensureExists(
    getHiddenLocalTracksByPid(state).get(pid),
    'Unable to get the hidden tracks from the given pid'
  );

/**
 * This selector gets the local track order for a PID, and ensures that one is
 * selected correctly. This makes it easier to avoid doing null checks everywhere.
 */
export const getLocalTrackOrder: DangerousSelectorWithArguments<
  TrackIndex[],
  Pid,
> = (state, pid) =>
  ensureExists(
    getLocalTrackOrderByPid(state).get(pid),
    'Unable to get the track order from the given pid'
  );

/**
 * Search strings filter a thread to only samples that match the strings.
 */
export const getSearchStrings: Selector<string[] | null> = createSelector(
  getCurrentSearchString,
  splitSearchString
);

export const getMarkersSearchStrings: Selector<string[] | null> =
  createSelector(getMarkersSearchString, splitSearchString);

export const getNetworkSearchStrings: Selector<string[] | null> =
  createSelector(getNetworkSearchString, splitSearchString);

/**
 * A RegExp can be used for searching and filtering the thread's samples.
 */
export const getSearchStringsAsRegExp: Selector<RegExp | null> = createSelector(
  getSearchStrings,
  stringsToRegExp
);

export const getMarkersSearchStringsAsRegExp: Selector<MarkerRegExps | null> =
  createSelector(getMarkersSearchStrings, stringsToMarkerRegExps);

export const getNetworkSearchStringsAsRegExp: Selector<MarkerRegExps | null> =
  createSelector(getNetworkSearchStrings, stringsToMarkerRegExps);

// Pre-allocate an array to help with strict equality tests in the selectors.
const EMPTY_TRANSFORM_STACK = [];

export const getTransformStack: DangerousSelectorWithArguments<
  TransformStack,
  ThreadsKey,
> = (state, threadsKey) => {
  return (
    getProfileSpecificState(state).transforms[threadsKey] ||
    EMPTY_TRANSFORM_STACK
  );
};

export const getIsBottomBoxOpen: Selector<boolean> = (state) => {
  const tab = getSelectedTab(state);
  return getProfileSpecificState(state).isBottomBoxOpenPerPanel[tab];
};

/**
 * The URL predictor is used to generate a link for an uploaded profile, to predict
 * what the URL will be.
 */
export const getUrlPredictor: Selector<(Action | Action[]) => string> =
  createSelector(
    getUrlState,
    (oldUrlState: UrlState) => (actionOrActionList: Action | Action[]) => {
      const actionList: Action[] = Array.isArray(actionOrActionList)
        ? actionOrActionList
        : [actionOrActionList];
      const newUrlState = actionList.reduce(urlStateReducer, oldUrlState);
      return urlFromState(newUrlState);
    }
  );

/**
 * Get the current path for a zip file that is being used.
 */
export const getPathInZipFileFromUrl: Selector<string | null> = (state) =>
  getUrlState(state).pathInZipFile;

/**
 * Get a short formatted string that represents the meta info of the current profile.
 */
export const getFormattedMetaInfoString: Selector<string | null> = (state) => {
  // Avoid circular dependencies by selected the profile meta manually.
  const { profile } = state.profileView;
  if (!profile) {
    return null;
  }

  return formatMetaInfoString(profile.meta);
};

/**
 * Get just the file name from the zip file path, if it exists.
 */
export const getFileNameInZipFilePath: Selector<string | null> = createSelector(
  getPathInZipFileFromUrl,
  (pathInZipFile) => {
    if (pathInZipFile) {
      return getProfileNameFromZipPath(pathInZipFile);
    }
    return null;
  }
);

/**
 * Get a profile name that can be used as a short identifier for the profile. This
 * will be displayed in the UI, and will be used as a default for a profile name
 * if none is currently set. If none is set, then a series of strategies will be
 * used to select a default one.
 */
export const getProfileNameWithDefault: Selector<string> = createSelector(
  getProfileNameFromUrl,
  getFileNameInZipFilePath,
  getFormattedMetaInfoString,
  (profileNameFromUrl, fileNameInZipFilePath, formattedMetaInfoString) => {
    // Always prefer a manually set name.
    if (profileNameFromUrl) {
      return profileNameFromUrl;
    }
    // Next, try and use a path from the zip file.
    if (fileNameInZipFilePath) {
      return fileNameInZipFilePath;
    }

    // Finally return a generic string describing the type of profile.
    if (!formattedMetaInfoString) {
      return 'Untitled profile';
    }
    return formattedMetaInfoString;
  }
);

/**
 * Determines the profile name used to store in the IndexedDB by default.
 */
export const getProfileNameForStorage: Selector<string> = createSelector(
  getProfileNameFromUrl,
  getFileNameInZipFilePath,
  (profileNameFromUrl, fileNameInZipFilePath) => {
    // Always prefer a manually set name.
    if (profileNameFromUrl) {
      return profileNameFromUrl;
    }
    // Next, try and use a path from the zip file.
    if (fileNameInZipFilePath) {
      return fileNameInZipFilePath;
    }

    // Finally, return a blank string.
    return '';
  }
);

function _shouldAllowSymbolServerUrl(symbolServerUrl) {
  try {
    const url = new URL(symbolServerUrl);
    if (isLocalURL(url)) {
      return true;
    }

    // For non-localhost hosts, we require HTTPS.
    const otherAllowedHostnames = [
      'symbols.mozilla.org',
      'symbolication.services.mozilla.com',
    ];
    if (!otherAllowedHostnames.includes(url.hostname)) {
      console.error(
        `The symbol server URL was not in the list of allowed domains. Rejecting ${symbolServerUrl} and defaulting to ${SYMBOL_SERVER_URL}.`
      );
      return false;
    }
    if (url.protocol !== 'https:') {
      console.error(
        `HTTPS is required for non-localhost symbol servers. Rejecting ${symbolServerUrl} and defaulting to ${SYMBOL_SERVER_URL}.`
      );
      return false;
    }
    return true;
  } catch (error) {
    console.error(
      `The symbol server URL was not valid. Rejecting ${symbolServerUrl} and defaulting to ${SYMBOL_SERVER_URL}.`,
      error
    );
    return false;
  }
}

export const getSymbolServerUrl: Selector<string> = createSelector(
  getUrlState,
  ({ symbolServerUrl }) => {
    if (!symbolServerUrl || !_shouldAllowSymbolServerUrl(symbolServerUrl)) {
      return SYMBOL_SERVER_URL;
    }

    // Strip off the trailing slash, if present.
    if (symbolServerUrl.endsWith('/')) {
      return symbolServerUrl.slice(0, -1);
    }
    return symbolServerUrl;
  }
);
