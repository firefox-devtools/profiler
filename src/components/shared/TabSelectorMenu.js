/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import * as React from 'react';
import { MenuItem } from '@firefox-devtools/react-contextmenu';
import { Localized } from '@fluent/react';
import classNames from 'classnames';

import { ContextMenu } from './ContextMenu';
import explicitConnect from 'firefox-profiler/utils/connect';
import { changeTabFilter } from 'firefox-profiler/actions/receive-profile';
import { getTabFilter } from '../../selectors/url-state';
import { getProfileFilterSortedPageData } from 'firefox-profiler/selectors/profile';
import { Icon } from 'firefox-profiler/components/shared/Icon';

import type { TabID, SortedTabPageData } from 'firefox-profiler/types';
import type { ConnectedProps } from 'firefox-profiler/utils/connect';

type StateProps = {
  +tabFilter: TabID | null,
  +sortedPageData: SortedTabPageData | null,
};

type DispatchProps = {
  +changeTabFilter: typeof changeTabFilter,
};

type Props = ConnectedProps<{}, StateProps, DispatchProps>;

import './TabSelectorMenu.css';

class TabSelectorMenuImpl extends React.PureComponent<Props> {
  _handleClick = (_event: SyntheticEvent<>, data: { id: TabID }): void => {
    this.props.changeTabFilter(data.id);
  };

  renderTabSelectorMenuContents() {
    const { sortedPageData, tabFilter } = this.props;
    if (!sortedPageData || sortedPageData.length === 0) {
      return null;
    }

    return (
      <>
        <MenuItem
          onClick={this._handleClick}
          data={{ id: null }}
          attributes={{
            className: classNames('tabSelectorMenuItem', {
              checkable: true,
              checked: tabFilter === null,
            }),
            'aria-checked': tabFilter === null ? 'false' : 'true',
          }}
        >
          <Localized id="TabSelectorMenu--all-tabs-and-windows">
            All tabs and windows
          </Localized>
        </MenuItem>
        {sortedPageData.map(({ tabID, pageData }) => (
          <MenuItem
            key={tabID}
            onClick={this._handleClick}
            data={{ id: tabID }}
            attributes={{
              className: classNames('tabSelectorMenuItem', {
                checkable: true,
                checked: tabFilter === tabID,
              }),
              'aria-checked': tabFilter === tabID ? 'false' : 'true',
            }}
          >
            <Icon iconUrl={pageData.favicon} />
            {pageData.hostname}
          </MenuItem>
        ))}
      </>
    );
  }

  render() {
    return (
      <ContextMenu id="TabSelectorMenu" className="TabSelectorMenu">
        {this.renderTabSelectorMenuContents()}
      </ContextMenu>
    );
  }
}

export const TabSelectorMenu = explicitConnect<{}, StateProps, DispatchProps>({
  mapStateToProps: (state) => ({
    tabFilter: getTabFilter(state),
    sortedPageData: getProfileFilterSortedPageData(state),
  }),
  mapDispatchToProps: {
    changeTabFilter,
  },
  component: TabSelectorMenuImpl,
});
