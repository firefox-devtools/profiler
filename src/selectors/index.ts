/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
export * from './app';
export * from './l10n';
export * from './per-thread';
export * from './profile';
export * from './url-state';
export * from './icons';
export * from './publish';
export * from './zipped-profiles';
export * from './cpu';
export * from './code';

import * as app from './app';
import {
  selectedThreadSelectors as selectedThread,
  selectedNodeSelectors as selectedNode,
} from './per-thread';
import * as profile from './profile';
import * as urlState from './url-state';
import * as icons from './icons';
import * as publish from './publish';
import * as zippedProfiles from './zipped-profiles';
import * as l10n from './l10n';
import * as cpu from './cpu';
import * as code from './code';

const _selectorsForConsole = {
  app,
  profile,
  urlState,
  icons,
  publish,
  zippedProfiles,
  selectedThread,
  selectedNode,
  l10n,
  cpu,
  code,
};

// Exports require explicit typing. Deduce the type with typeof.
export const selectorsForConsole: typeof _selectorsForConsole =
  _selectorsForConsole;
