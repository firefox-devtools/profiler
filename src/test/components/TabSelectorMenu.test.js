/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import React from 'react';
import { Provider } from 'react-redux';
import { screen } from '@testing-library/react';

import { render } from 'firefox-profiler/test/fixtures/testing-library';
import { TabSelectorMenu } from 'firefox-profiler/components/shared/TabSelectorMenu';
import { addActiveTabInformationToProfile } from '../fixtures/profiles/processed-profile';
import {
  getProfileWithNiceTracks,
  getHumanReadableTracks,
} from '../fixtures/profiles/tracks';
import { storeWithProfile } from '../fixtures/stores';
import { fireFullClick } from '../fixtures/utils';
import { getTabFilter } from '../../selectors/url-state';
import { ensureExists } from 'firefox-profiler/utils/flow';
import { removeURLs } from 'firefox-profiler/utils/string';

describe('app/TabSelectorMenu', () => {
  function setup() {
    const { profile, ...extraPageData } = addActiveTabInformationToProfile(
      getProfileWithNiceTracks()
    );
    ensureExists(profile.pages)[3].favicon =
      'data:image/png;base64,test-png-favicon-data-for-profiler.firefox.com';

    // This is needed for the thread activity score calculation.
    profile.meta.sampleUnits = {
      time: 'ms',
      eventDelay: 'ms',
      threadCPUDelta: 'ns',
    };

    // Add some frames with innerWindowIDs now. Note that we only expand the
    // innerWindowID array and not the others as we don't check them at all.
    //
    // Thread 0 will be present in firstTabTabID.
    // Thread 1 be present in secondTabTabID.
    profile.threads[0].frameTable.innerWindowID[0] =
      extraPageData.parentInnerWindowIDsWithChildren;
    profile.threads[0].frameTable.length++;

    // Add a threadCPUDelta value for thread activity score.
    profile.threads[0].samples.threadCPUDelta = [1];

    profile.threads[1].frameTable.innerWindowID[0] =
      extraPageData.secondTabInnerWindowIDs[0];
    profile.threads[1].frameTable.length++;
    // Add a threadCPUDelta value for thread activity score. This thread
    // should stay above the first thread.
    profile.threads[0].samples.threadCPUDelta = [2];

    const store = storeWithProfile(profile);
    render(
      <Provider store={store}>
        <TabSelectorMenu />
      </Provider>
    );

    return {
      profile,
      ...extraPageData,
      ...store,
    };
  }

  it('should render properly', () => {
    setup();
    expect(document.body).toMatchSnapshot();
  });

  it('should not render when the profile does not contain any page data', () => {
    const store = storeWithProfile(getProfileWithNiceTracks());
    render(
      <Provider store={store}>
        <TabSelectorMenu />
      </Provider>
    );
    expect(document.body).toMatchSnapshot();
  });

  it('should switch tabs properly', () => {
    const { getState, firstTabTabID, secondTabTabID } = setup();

    // Check that there is no tab filter at first.
    expect(getTabFilter(getState())).toBe(null);

    // Change the tab filter by clicking on the menu item.
    const mozillaTab = screen.getByText('mozilla.org');
    fireFullClick(mozillaTab);

    // Check the tab filter again, it should match the first tab in the profile.
    expect(getTabFilter(getState())).toBe(firstTabTabID);

    // Change the tab filter again.
    const profilerTab = screen.getByText('profiler.firefox.com');
    fireFullClick(profilerTab);

    // Check the tab filter again, it should match the second tab in the profile.
    expect(getTabFilter(getState())).toBe(secondTabTabID);

    // Change the tab filter to all tabs and windows
    const allTabs = screen.getByText('All tabs and windows');
    fireFullClick(allTabs);

    // Check the tab filter again, it should be null, meaning all tabs and windows.
    expect(getTabFilter(getState())).toBe(null);
  });

  it('should display the relevant threads after tab switch', () => {
    const { getState, firstTabTabID, secondTabTabID } = setup();

    // Check that there is no tab filter at first.
    expect(getTabFilter(getState())).toBe(null);
    // Also make sure that we have all the threads currently.
    expect(getHumanReadableTracks(getState())).toEqual([
      'show [thread GeckoMain default]',
      'show [thread GeckoMain tab] SELECTED',
      '  - show [thread DOM Worker]',
      '  - show [thread Style]',
    ]);

    // Change the tab filter by clicking on the menu item.
    const profilerTab = screen.getByText('profiler.firefox.com');
    fireFullClick(profilerTab);

    // Check the tab filter again, it should match the second tab in the profile.
    expect(getTabFilter(getState())).toBe(secondTabTabID);
    // Make sure that the second process group is visible.
    // Note that the first thread will be visible too, because it's the parent
    // process which we always include.
    expect(getHumanReadableTracks(getState())).toEqual([
      'hide [thread GeckoMain default]',
      'show [thread GeckoMain tab] SELECTED',
      '  - show [thread DOM Worker]',
      '  - show [thread Style]',
    ]);

    // Change the tab filter again.
    const mozillaTab = screen.getByText('mozilla.org');
    fireFullClick(mozillaTab);

    // Check the tab filter again, it should match the first tab in the profile.
    expect(getTabFilter(getState())).toBe(firstTabTabID);
    // Also make sure that the first process is visible.
    expect(getHumanReadableTracks(getState())).toEqual([
      'show [thread GeckoMain default] SELECTED',
    ]);

    // Change the tab filter to all tabs and windows.
    const allTabs = screen.getByText('All tabs and windows');
    fireFullClick(allTabs);

    // Check the tab filter again, it should be null, meaning full profile.
    expect(getTabFilter(getState())).toBe(null);
    // It should show the full thread list again.
    expect(getHumanReadableTracks(getState())).toEqual([
      'show [thread GeckoMain default]',
      'show [thread GeckoMain tab] SELECTED',
      '  - show [thread DOM Worker]',
      '  - show [thread Style]',
    ]);
  });

  it('should sort the tabs by their activity scores', () => {
    setup();

    const profilerTab = screen.getByText('profiler.firefox.com');
    const mozillaTab = screen.getByText('mozilla.org');

    // Make sure that profiler tab comes before the mozilla tab.
    expect(profilerTab.compareDocumentPosition(mozillaTab)).toBe(
      Node.DOCUMENT_POSITION_FOLLOWING
    );
  });

  it('should render sanitized page urls correctly', () => {
    const { profile, ...extraPageData } = addActiveTabInformationToProfile(
      getProfileWithNiceTracks()
    );
    // This is needed for the thread activity score calculation.
    profile.meta.sampleUnits = {
      time: 'ms',
      eventDelay: 'ms',
      threadCPUDelta: 'ns',
    };

    // Add a webextension url to test it.
    ensureExists(profile.pages)[4].url =
      'moz-extension://259ec0ce-9df7-8e4a-ad30-3b67bed900f3/';

    // Sanitize the page urls.
    profile.pages = ensureExists(profile.pages).map((page, index) => ({
      ...page,
      url: removeURLs(page.url, `<Page #${index}>`),
    }));

    // Attach innerWindowIDs to the samples.
    profile.threads[0].frameTable.innerWindowID[0] =
      extraPageData.parentInnerWindowIDsWithChildren;
    profile.threads[0].frameTable.length++;
    profile.threads[0].frameTable.innerWindowID[1] =
      extraPageData.secondTabInnerWindowIDs[0];
    profile.threads[0].frameTable.length++;

    const store = storeWithProfile(profile);
    render(
      <Provider store={store}>
        <TabSelectorMenu />
      </Provider>
    );

    // Make sure that sanitized https and moz-extension urls are still visible.
    expect(screen.getByText('https://', { exact: false })).toBeInTheDocument();
    expect(
      screen.getByText('moz-extension://', { exact: false })
    ).toBeInTheDocument();
    expect(document.body).toMatchSnapshot();
  });
});
