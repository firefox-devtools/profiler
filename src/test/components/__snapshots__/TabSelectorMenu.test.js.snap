// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app/TabSelectorMenu should not render when the profile does not contain any page data 1`] = `
<body>
  <div>
    <nav
      class="react-contextmenu TabSelectorMenu"
      role="menu"
      style="position: fixed; opacity: 0; pointer-events: none;"
      tabindex="-1"
    >
      <div />
    </nav>
  </div>
</body>
`;

exports[`app/TabSelectorMenu should render properly 1`] = `
<body>
  <div>
    <nav
      class="react-contextmenu TabSelectorMenu"
      role="menu"
      style="position: fixed; opacity: 0; pointer-events: none;"
      tabindex="-1"
    >
      <div
        aria-checked="false"
        aria-disabled="false"
        class="react-contextmenu-item tabSelectorMenuItem checkable checked"
        role="menuitem"
        tabindex="-1"
      >
        All tabs and windows
      </div>
      <div
        aria-checked="true"
        aria-disabled="false"
        class="react-contextmenu-item tabSelectorMenuItem checkable"
        role="menuitem"
        tabindex="-1"
      >
        <div
          class="nodeIcon "
        />
        profiler.firefox.com
      </div>
      <div
        aria-checked="true"
        aria-disabled="false"
        class="react-contextmenu-item tabSelectorMenuItem checkable"
        role="menuitem"
        tabindex="-1"
      >
        <div
          class="nodeIcon "
        />
        mozilla.org
      </div>
    </nav>
  </div>
</body>
`;

exports[`app/TabSelectorMenu should render sanitized page urls correctly 1`] = `
<body>
  <div>
    <nav
      class="react-contextmenu TabSelectorMenu"
      role="menu"
      style="position: fixed; opacity: 0; pointer-events: none;"
      tabindex="-1"
    >
      <div
        aria-checked="false"
        aria-disabled="false"
        class="react-contextmenu-item tabSelectorMenuItem checkable checked"
        role="menuitem"
        tabindex="-1"
      >
        All tabs and windows
      </div>
      <div
        aria-checked="true"
        aria-disabled="false"
        class="react-contextmenu-item tabSelectorMenuItem checkable"
        role="menuitem"
        tabindex="-1"
      >
        <div
          class="nodeIcon "
        />
        moz-extension://&lt;Page #4&gt;
      </div>
      <div
        aria-checked="true"
        aria-disabled="false"
        class="react-contextmenu-item tabSelectorMenuItem checkable"
        role="menuitem"
        tabindex="-1"
      >
        <div
          class="nodeIcon "
        />
        https://&lt;Page #3&gt;
      </div>
    </nav>
  </div>
</body>
`;
